appinventor/blocklyeditor/src/blocks/logic.js

```diff
@@ -203,13 +195,34 @@ Blockly.Blocks['logic_operation'] = {
       }
     }
 
+    if (this.itemCount_ === 0) {
+      this.removeInput(this.emptyInputName, true);
+    }
+    if (this.itemCount_ > 0) {
+      this.removeInput('A', true);
+    }
+    if (this.itemCount_ > 1) {
+      this.removeInput('B', true);
+    }
     for (var x = 2; x < this.itemCount_; x++) {
-      this.removeInput(this.repeatingInputName + x);
+      this.removeInput(this.repeatingInputName + x, true);
     }
     this.itemCount_ = window.parseInt(container.getAttribute('items'), 10);
-    for (var x = 2; x < this.itemCount_; x++) {
+    for (var x = 0; x < this.itemCount_; x++) {
       this.addInput(x);
     }
+    if (this.itemCount_ === 0) {
+      this.addEmptyInput();
+    }
+    // NOTE(ewp): Blockly doesn't trigger the validation function when the field is set during
+    // load, so we override setValue here to make sure that the additional and/or labels (if
+    // present) match the dropdown's value.
```

Comment:
If you mean load as in loading from XML it actually does in the latest version. So you might want to add a TODO to remove this after the update =)

Reply:
Thanks for the info. Some day it would be nice if we could have the validation/update logic be separated but I expect that's unlikely to happen any time soon. In this particular case, the function is idempotent so it shouldn't matter, but less code executed is better.